tools/loader/ImageClassifier.cpp

```diff
@@ -457,34 +479,83 @@ int main(int argc, char **argv) {
-    GLOW_ASSERT(inputImagePH->dims() == inputImageData.dims() &&
+    GLOW_ASSERT(inputImagePH->dims() == inputImageData->dims() &&
                 "New input shape does not match the compiled function.");
 
     // Convert the raw input to fp16. This must be done every time we get new
     // image data.
     if (convertInAndOutToFp16) {
-      inputImageData.convertToType(ElemKind::Float16Ty);
+      inputImageData->convertToType(ElemKind::Float16Ty);
     }
 
     // About to run inference, so update the input image Placeholder's backing
     // Tensor with inputImageData.
-    updateInputPlaceholders(bindings, {inputImagePH}, {&inputImageData});
+    updateInputPlaceholders(*bindings, {inputImagePH}, {inputImageData.get()});
 
     // Perform the inference execution, updating SMT.
-    auto batchSize = inputImageData.dims()[0];
-    loader.runInference(bindings, batchSize);
-
-    // Print the top-k results from the output Softmax tensor.
-    numErrors += processAndPrintResults(SMT, inputImageBatchFilenames);
+    auto batchSize = inputImageData->dims()[0];
+    if (preProcessImages) {
+      auto newContext = llvm::make_unique<ExecutionContext>();
+      auto ph = newContext->getPlaceholderBindings();
+      ph->allocate(inputImagePH);
+      ph->allocate(outputPH);
+      updateInputPlaceholders(*ph, {inputImagePH}, {inputImageData.get()});
+
+      preLoadedImages.push_back(std::move(inputImageData));
+      contexts.emplace(inputImageBatchFilenames, std::move(newContext));
+    } else {
+      loader.runInference(*bindings, batchSize);
+      // Print the top-k results from the output Softmax tensor.
+      numErrors += processAndPrintResults(SMT, inputImageBatchFilenames);
+    }
+  }
+  if (preProcessImages) {
+    runtime::HostManager *hostManager = loader.getHostManager();
+    std::string name = loader.getFunctionName();
+    std::atomic<unsigned> inFlight(0);
+    std::promise<void> runPromise;
+    auto fut = runPromise.get_future();
+    if (timeOpt) {
+      timer.startTimer();
+    }
+    for (auto &batch : contexts) {
+      auto output = batch.second->getPlaceholderBindings()->get(outputPH);
+      auto names = batch.first;
+      llvm::Error runErr = llvm::Error::success();
+      inFlight++;
+      hostManager->runNetwork(
+          name, std::move(batch.second),
+          [&runErr, output, names, &runPromise,
+           &inFlight](runtime::RunIdentifierTy, llvm::Error err,
+                      std::unique_ptr<ExecutionContext> contextPtr) {
+            runErr = std::move(err);
+            // Print the top-k results from the output Softmax tensor.
+            processAndPrintResults(output, names);
+            if (--inFlight == 0) {
+              runPromise.set_value();
+            }
+          });
+
+      EXIT_ON_ERR(std::move(runErr));
+    }
+    // wait for all to finish.
+    fut.wait();
+    if (timeOpt) {
+      timer.stopTimer();
+      llvm::outs() << llvm::formatv("Wall time per item (s): {0:f4}\n",
```

Comment:
This PR also adds timing/benchmarking functionality to the ImageClassifier? Could you update the summary to reflect this (& how it works, what it looks like, etc)?